src/include/encoding.h

```diff
@@ -617,30 +617,18 @@ inline std::enable_if_t<!traits::supported>
 {
   __u32 n = (__u32)(ls.size());  // c++11 std::list::size() is O(1)
   encode(n, bl);
-  for (auto p = ls.begin(); p != ls.end(); ++p)
-    encode(*p, bl);
+  for (auto& elem : ls) {
+    encode(elem, bl);
+  }
 }
 template<class T, class Alloc, typename traits>
 inline std::enable_if_t<!traits::supported>
   encode(const std::list<T,Alloc>& ls, bufferlist& bl, uint64_t features)
 {
-  // should i pre- or post- count?
-  if (!ls.empty()) {
-    unsigned pos = bl.length();
-    unsigned n = 0;
-    encode(n, bl);
-    for (auto p = ls.begin(); p != ls.end(); ++p) {
-      n++;
-      encode(*p, bl, features);
-    }
-    ceph_le32 en;
-    en = n;
-    bl.copy_in(pos, sizeof(en), (char*)&en);
-  } else {
-    __u32 n = (__u32)(ls.size());    // FIXME: this is slow on a list.
-    encode(n, bl);
-    for (auto p = ls.begin(); p != ls.end(); ++p)
-      encode(*p, bl, features);
+  __u32 n = (__u32)(ls.size());  // c++11 std::list::size() is O(1)
```

Comment:
this could introduce performance regression. i'd suggest put something like
```c++
auto size = 0U;
encode(size, bl);
auto size_out = bl.begin();
size_out.advance(-sizeof(size));
for (auto& elem : ls) {
  encode(elem, bl, features);
  size++;
}
ceph_le32 en = size;
size_out.copy_in(sizeof(en), (char*)&en);
```

Reply:
Would it be the regression caused by something other than the buggy, non-C+11-compliant `std::list::size()`?

Reply:
yeah, i concern is just the O(n) `std::list::size()` from libstdc++ shipped by in SCLs' GCC-7. it tries to be ABI compatible with pre-C++11.

Reply:
Ouch, thanks for pointing this out. I thought the `::size()` issue is behind us. :-/ Well, DNM for now and let's wait on ultimately  judging the `append_hole` as it could be useful here.

Reply:
@tchaikov: this also means we need to take care about the check in `src/CMakeLists.txt`:

```
if(CMAKE_COMPILER_IS_GNUCXX AND
    CMAKE_CXX_COMPILER_VERSION VERSION_LESS 5.1)
  # this is not always correct, as one can use clang with libstdc++ or
  # use old gcc with new libstdc++, but it covers the most cases.
  #
  # libstdc++ 4.9 has O(n) list::size(), and its regex is buggy
  message(SEND_ERROR "performance regression is expected due to an O(n) "
    "implementation of 'std::list::size()' in libstdc++ older than 5.1.0, "
    "Please use GCC 5.1 and up.")
endif()
```

Reply:
yeah, i added that check. that predates when we moved to GCC 7.3. and i think

1. we already check for C++17 support in the same `CMakeLists.txt`
1. it's non trivial to check the O(1) `std::list::size()`.
1. we always build the GCC 7.3 from SCLs on CentOS and RHEL. so the error message is always printed if we could have detected the O(1) `std::list::size()`.

so probably i should remove it, as it basically useless. =)

Reply:
#24477